---
title: "Figures_Cobs3.1"
output:
  html_document:
    df_print: paged
---

This script to generate figures for the Cobs3.1 manuscript. Started working in this on the 26 Dec 2022
```{r warning=FALSE}
# load all needed libraries:

library(readr)
library(dplyr)
library(tidyverse)
library(data.table)
library(reshape2)
library(ggrepel)
library(ggpubr)
library(ggbeeswarm)
library(circlize)
library(grid)

```

### analyze and plot correlations based on LM
Estimation of recombination rate per chromosome (only covered parts with markers)

```{r}
# load LM data
lg <- read_delim("datasets/manuallyCurated.markers.forR.txt", col_names = F, delim = "\t")
colnames(lg) <- c("chr", "pos", "id", "distance1", "lg", "distance")

# load chrom size data
scf = read.table("datasets/Cobs3.1.chrom-size.txt")
scf = scf[seq(1,29,1),]
names(scf) = c("chr", "fullphyslength")

# reorder lg:
plotthis <- as.data.frame(lg[,c(5,2,6,3)])
colnames(plotthis) <- c("lg", "physical", "genetic", "locus")

# get size of the map
mapPhySize <- aggregate(plotthis$genetic, by = list(plotthis$lg), max)

# get physical size of regions with markers
mapPhySize$physLength <- aggregate(physical ~ lg, plotthis, FUN = function(i)max(i) - min(i))[,2]
colnames(mapPhySize) <- c("chr", "mapLength", "physLength")

# add actual size of LGs:
mapPhySize <- merge(mapPhySize, scf, by = "chr")

# get actual recombination rates per chromosome
mapPhySize$recomRate <- mapPhySize$mapLength/(mapPhySize$physLength * 1E-6)



# get no of markers per LG:
df2 <- plotthis %>% group_by(lg) %>% summarise(length(physical))
df2 <- df2[order(df2$`length(physical)`, decreasing = T),]
names(df2) = c("chr", "markers")


# combine the two:
mapPhySize <- merge(mapPhySize, df2, by = c("chr"))
mapPhySize

# add marker density:
mapPhySize$density <- mapPhySize$markers/(mapPhySize$physLength/250000)

# get Table 1:

table1 <- mapPhySize[,1:5]
```

```{r}
# additional metrics:

## Marker density:
mean_ci(mapPhySize$density)

## fraction covered with markers = 162.866 Mb
sum(mapPhySize$physLength)* 1E-6 


## calculate mean/ci per chromosome group
###(< 5 Mb):
mean_ci(mapPhySize[mapPhySize$fullphyslength < 5E6, ]$recomRate)
# mean    ymin      ymax
# 13.29613	8.523973	18.06829	

### (> 10 Mb):
mean_ci(mapPhySize[mapPhySize$fullphyslength > 10E6, ]$recomRate)
# mean    ymin      ymax
# 6.967698	3.647343	10.28805

### (>= 5 Mb& <=10 Mb):
mean_ci(mapPhySize[mapPhySize$fullphyslength >= 5E6 & mapPhySize$fullphyslength <= 10E6, ]$recomRate)
# mean    ymin      ymax
# 7.072861	4.594686	9.551036		
```

Calculate recombination rates with no filters:
```{r}
# get average recombination rate considering fraction covered with markers: 8.1 cM/Mb
sum(mapPhySize$mapLength)/(sum(mapPhySize$physLength)* 1E-6)
```

```{r}
# The fraction of the genome covered by enough markers is: 82.8 %
genome <- 193051128 * 1E-6
gp <- (3846852 - 799255) * 1E-6 # for the gap on LG11
farctionCovered <- (sum(mapPhySize$physLength)* 1E-6) - gp

farctionCovered/genome * 100
```

### Make Circos plot:

Load and prepare data:
```{r}
# load genome
ref <- read_delim("datasets/Cobs3.1.chrom-size.txt", col_names = F)
ref <- ref %>% filter(ref$X2>1e6)
ref$X1 = as.numeric(gsub("LG", "", ref$X1))
ref <- ref[ref$X1 %in% c(1:29),]


### load gold
gold <- read_delim("datasets/final.metrics.in250kbwindows.tsv", col_names = T, delim = "\t")


### RR
rr <- gold[,c(1:4)]
colnames(rr) <- c("chr", "start", "end", "value1")
rr <- rr %>%
  filter(str_detect(chr, "LG"))

rr <- rr[with(rr, order(chr, start)), ]
rr$chr = as.numeric(gsub("LG", "", rr$chr))
rr <- rr[rr$chr %in% c(1:29),]


### load TE
te <- gold[,c(1:3,5)]
colnames(te) <- c("chr", "start", "end", "value1")
te <- te %>%
  filter(str_detect(chr, "LG"))

te <- te[with(te, order(chr, start)), ]
te$chr = as.numeric(gsub("LG", "", te$chr))
te <- te[te$chr %in% c(1:29),]


### load TEI
teisl <- read_delim("datasets/TEislands.w250kb.bed", col_names = F)
colnames(teisl) <- c("chr", "start", "end")
teisl <- teisl %>%
  filter(str_detect(chr, "LG"))
teisl$chr <- as.numeric(sub("LG", "", teisl$chr))
teisl <- teisl[teisl$chr %in% c(1:29),]


### load gene
gene <- gold[,c(1:3,8)]
colnames(gene) <- c("chr", "start", "end", "value1")
gene <- gene %>%
  filter(str_detect(chr, "LG"))

gene <- gene[with(gene, order(chr, start)), ]
gene$chr = as.numeric(gsub("LG", "", gene$chr))
gene <- gene[gene$chr %in% c(1:29),]


### load sv
sv <- read_delim("datasets/SVs/sv.bed.filtered.repeat.0.2.txt", col_names = T)
sv$end <- sv$end+50000
sv <- sv[,c(1:3)]
sv <- sv %>%
  filter(str_detect(chr, "LG"))
sv$chr <- as.numeric(sub("LG", "", sv$chr))
sv <- sv[sv$chr %in% c(1:29),]



### load introgressed regions
intro <- read_delim("datasets/introgression.250kb.bed", col_names = F)
colnames(intro) <- c("chr", "start", "end")
intro <- intro %>%
  filter(str_detect(chr, "LG"))
intro$chr <- as.numeric(sub("LG", "", intro$chr))
intro <- intro[intro$chr %in% c(1:29),]


### load pi Itabuna
pii <- gold[,c(1:3,11)]
pii <- pii %>%
  filter(str_detect(chr, "LG"))

colnames(pii) <- c("chr", "start", "end", "value1")
pii <- pii[with(pii, order(chr, start)), ]
pii$chr = as.numeric(gsub("LG", "", pii$chr))
pii <- pii[pii$chr %in% c(1:29),]



####  load pi Una
piu <- gold[,c(1:3,10)]
piu <- piu %>%
  filter(str_detect(chr, "LG"))

colnames(piu) <- c("chr", "start", "end", "value1")
piu <- piu[with(piu, order(chr, start)), ]
piu$chr = as.numeric(gsub("LG", "", piu$chr))
piu <- piu[piu$chr %in% c(1:29),]
```

Make the circos plot layer by layer
```{r}
circos.clear()
col_text <- "black"
circos.par("track.height"=0.4, gap.degree=c(.2, .2, .2, .2, .2,
                                            .2, .2, .2, .2, .2,
                                            .2, .2, .2, .2, .2,
                                            .2, .2, .2, .2, .2,
                                            .2, .2, .2, .2, .2,
                                            .2, .2, .2, 6), cell.padding=c(0, 0, 0, 0), start.degree = 90, track.margin =c(0.01, 0.01))

circos.initialize(factors=c("1","2","3","4","5","6","7","8",
                            "9","10","11","12","13","14","15",
                            "16","17","18","19","20","21","22",
                            "23","24","25","26","27","28", "29"),
                  xlim=matrix(c(rep(0, 29), ref$X2), ncol=2))



circos.info()

# add genome
circos.track(ylim=c(0, 1), panel.fun=function(x, y) {
  chr=CELL_META$sector.index
  xlim=CELL_META$xlim
  ylim=CELL_META$ylim
  circos.text(mean(xlim), mean(ylim), chr, cex=0.35, col=col_text,
              facing="bending.inside", niceFacing=TRUE)
}, bg.col="grey95", bg.border=T, track.height=0.04, bg.lwd = 0.2)


# add genome x axis
brk <- c(0, 2, 4, 6, 8, 10, 12, 14)*10^6
circos.track(track.index = get.current.track.index(), panel.fun=function(x, y) {
  circos.axis(h="top", major.at=brk, labels=round(brk/10^6, 1), labels.cex=0.35,
              col=col_text, labels.col=col_text, lwd=0.7, labels.facing="clockwise", minor.ticks = F)
}, bg.border=F)


# Add SVs positions:
data.TT <- sv
data.TT[,2] <- as.numeric(unlist(data.TT[,2]))
data.TT[,3] <- as.numeric(unlist(data.TT[,3]))
data.TT$num <- 1:nrow(data.TT)
rectcols <- c("#226C3B")
data.TT[,4] <- rectcols
tktransparency <- 1
circos.genomicTrackPlotRegion(data.TT, ylim=c(0,1), track.margin = c(0.01,0.0001), track.height =0.01,
                              bg.col = c("NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA",
                                         "NA","NA","NA","NA","NA","NA","NA","NA","NA"), 
                              bg.border = NA, panel.fun = function(region,value,...){
                                circos.genomicRect(region, value, col=adjustcolor(value[[1]],alpha.f = tktransparency), border = NA, ...)	  
                              })



# Add the line plot track for recombination rate
circos.track(factors=rr$chr, x=rr$start, y=rr$value1, track.margin = c(0.005,0.01), panel.fun=function(x, y) {
  circos.segments(x0=0, x1=get.cell.meta.data("xlim"), y0=30, y1=30, lwd=0.1, col="grey")
  circos.segments(x0=0, x1=get.cell.meta.data("xlim"), y0=60, y1=60, lwd=0.1, col="grey")
  circos.lines(x, y, col="#3CACAE", lwd=1.5, area = F, type="l",border = "#3CACAE")
  circos.barplot(pos = x, value = y, border = "#3CACAE", col="#3CACAE", lwd = 1.75)
}, ylim=range(rr$value1, na.rm = T), track.height=0.12, bg.border=T, bg.lwd = 0.2)
circos.yaxis(at=c(30, 60), sector.index = 1, labels.cex=.4, lwd=0, tick.length=0, labels.col=col_text, col="#FFFFFF")


# Add introgression regions
data.TT <- intro
data.TT[,2] <- as.numeric(unlist(data.TT[,2]))
data.TT[,3] <- as.numeric(unlist(data.TT[,3]))
data.TT$num <- 1:nrow(data.TT)
rectcols <- c("#6C0BA9")
data.TT[,4] <- rectcols
tktransparency <- 1
circos.genomicTrackPlotRegion(data.TT, ylim=c(0,1), track.margin = c(0.01,0), track.height =0.01,
                              bg.col = c("NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA",
                                         "NA","NA","NA","NA","NA","NA","NA","NA","NA"), 
                              bg.border = NA, panel.fun = function(region,value,...){
                                circos.genomicRect(region, value, col=adjustcolor(value[[1]],alpha.f = tktransparency), border = NA, ...)	  
                              })



# Add the heatmap track for TE content
circos.genomicHeatmap(te, col = colorRamp2(c(0, 0.08333333, 0.16666666, 0.24999999, 0.33333332, 0.41666665, 0.49999998, 0.58333331,
                                             0.66666664, 0.74999997, 0.83333330, 0.91666663, 0.99999996), 
                                           c("#0075dc", "#009ef9", "#00c0ee", "#00dcbf", "#00f27e", "#81f256",
                                             "#bdef2f", "#f2e800", "#f8d700", "#fcc600", "#ffb500", "#ffa405","orange")),
                      track.margin = c(0.01, 0.01), connection_height = NULL, heatmap_height = 0.03,border = NA)



# Add TE islands
data.TT <- teisl
data.TT[,2] <- as.numeric(unlist(data.TT[,2]))
data.TT[,3] <- as.numeric(unlist(data.TT[,3]))
data.TT$num <- 1:nrow(data.TT)
rectcols <- c("black")
data.TT[,4] <- rectcols
tktransparency <- 1
circos.genomicTrackPlotRegion(data.TT, ylim=c(0,1), track.margin = c(0.0001,0), track.height =0.005,
                              bg.col = c("NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA",
                                         "NA","NA","NA","NA","NA","NA","NA","NA","NA"),
                              bg.border = NA, panel.fun = function(region,value,...){
                                circos.genomicRect(region, value, col=adjustcolor(value[[1]],alpha.f = tktransparency), border = NA, ...)
                              })




# Add the areaplot track for gene content
circos.track(factors=gene$chr, x=(gene$start+gene$end)*0.5, y=gene$value1, panel.fun=function(x, y) {
  circos.segments(x0=0, x1=get.cell.meta.data("xlim"), y0=0.2, y1=0.2, lwd=0.1, col="grey")
  circos.segments(x0=0, x1=get.cell.meta.data("xlim"), y0=0.4, y1=0.4, lwd=0.1, col="grey")
  # circos.segments(x0=0, x1=get.cell.meta.data("xlim"), y0=0.6, y1=0.6, lwd=0.5, col="grey")
  circos.barplot(pos = x, value = y, border = "limegreen", col="limegreen", lwd = 1.5)
}, ylim=range(gene$value1), track.height=.1, bg.border=T, bg.lwd = 0.2)
# gene y axis
circos.yaxis(at=c(0.2, 0.4), sector.index = 1, labels.cex=.4, lwd=0, tick.length=0, labels.col=col_text, col="#FFFFFF")


# Add the lineplot track for nucleotide diversity
# Pi Itabuna
Itabuna <- scales::alpha("#B79F00", alpha=0.6)

circos.track(factors=pii$chr, x=pii$start, y=log10(pii$value1), panel.fun=function(x, y) {
  circos.segments(x0=0, x1=get.cell.meta.data("xlim"), y0=-5, y1=-5, lwd=0.1, col="grey")
  circos.segments(x0=0, x1=get.cell.meta.data("xlim"), y0=-3, y1=-3, lwd=0.1, col="grey")
  circos.segments(x0=0, x1=get.cell.meta.data("xlim"), y0=-4, y1=-4, lwd=0.1, col="grey")
  circos.lines(x, y, col=Itabuna, lwd=1, area = F)
}, ylim=range(c(log10(pii$value1),log10(piu$value1)), na.rm = T), track.height=0.12, bg.border=T, bg.lwd = 0.2)
# pi y axis
circos.yaxis(at=c( -5, -3, -4), sector.index = 1, labels.cex=.4, lwd=0, tick.length=0, labels.col=col_text, col="#FFFFFF")

## add extra lines pi
# Pi Una
Una <- scales::alpha("#821D30", alpha=0.6)

circos.track(factors=piu$chr, x=piu$start, y=log10(piu$value1), track.index = 8, panel.fun=function(x, y) {
  circos.lines(x, y, col=Una,  lwd=1, area = F)
}, ylim=range(log10(piu$value1), na.rm = T), track.height=0.12, bg.border=F)

```

Make boxplots accompaining the circos plot. 
RR in TE isl vs LDRs
```{r}
# load data:
gold <- read_delim("datasets/final.metrics.in250kbwindows.tsv", col_names = T, delim = "\t")
gold <- gold[,c(1:7,26)]

# mask gap on LG11
gold[gold$CHR == 11 & gold$start >= 750000 & gold$end <=4000000,]$RR <- NA


head(gold)
```

```{r}
rrTE <- gold %>%ggplot(.,aes(x = region, y = RR, fill = region, color=region))+
  geom_quasirandom(alpha=.7,size=.3,pch=21,aes(col=region,fill=region), width = 1/8)+
  geom_boxplot(aes(fill=region),alpha=.1,outlier.colour = NA,width=.25,lwd=.1,color="black")+
  scale_color_manual(values=c("#0075DC","#FFA405"))+
  scale_fill_manual(values=c( "#0075DC","#FFA405"))+
  stat_compare_means(aes(label = ..p.signif..), label.x = 1.4, label.y = 70)+
  theme_bw()+
  xlab("")+ #ylim(c(0,0.005))+
  scale_x_discrete(limits=c("LDRs","TE islands"))+
  ylab(expression(paste("recombination rate (cM/Mb)")))+
  theme(axis.text=element_text(size=7),
        axis.title=element_text(size=8,face="bold"),
        axis.text.x = element_text(angle = 45, hjust = 1),
        legend.position = "none",
        panel.border = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        plot.title = element_text(size = 9),
        axis.line = element_line(colour = "black"))+
  ggtitle("(b)")


```

RR in LDRs introgressed vs LDRs others

```{r}
dt <- gold[gold$CHR %in% c(2,3,
                            5,7,9,10,
                            11,13,14,15,
                            16,17,18,19,20,
                            21,22,23,24,
                            25,26,27,28,29)  & gold$region == "LDRs",]

rrIntr <- dt %>%ggplot(.,aes(x = introgression, y = RR, fill = introgression, color=introgression))+
  geom_quasirandom(alpha=.7,size=.3,pch=21,aes(col=introgression,fill=introgression), width = 1/8)+
  geom_boxplot(aes(fill=introgression),alpha=.1,outlier.colour = NA,width=.25,lwd=.1,color="black")+
  scale_color_manual(values=c("#0075DC", "#6C0BA9"))+
  scale_fill_manual(values=c("#0075DC", "#6C0BA9"))+
  stat_compare_means(aes(label = ..p.signif..), label.x = 1.4, label.y = 70)+
  theme_bw()+
  xlab("")+ #ylim(c(0,0.005))+
  ylab(expression(paste("recombination rate (cM/Mb)")))+
  scale_x_discrete(limits=c("no", "yes"), labels = c("LDRs other", "LDRs introgr."))+
  theme(axis.text=element_text(size=7),
        axis.title=element_text(size=8,face="bold"),
        axis.text.x = element_text(angle = 45, hjust = 1),
        axis.title.y = element_blank(),
        axis.text.y = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.y = element_blank(),
        legend.position = "none",
        panel.border = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        plot.title = element_text(size = 9),
        axis.line = element_line(colour = "black"))+
  ggtitle("(c)")

```


diversity in LDRs recombining vs LDRs nonrecombining

```{r}
# load data:
gold <- read_delim("datasets/final.metrics.in250kbwindows.tsv", col_names = T, delim = "\t")
gold <- gold[,c(1:7,10:13,26)]

# filter LGs with large intrgeressions and or having very few localized markers:
data <- subset(gold, !(gold$CHR %in% c(9,10,11,17,21,26,29)))
data <- data[data$introgression == "no",]

# Exclude LDRs for diversity
data <- data[data$region == "LDRs",]


# define candidates
data <- data[!is.na(data$RR),]

data <- mutate(data, candidate = ifelse(data$CHR == 1 & data$RR == 0|
                                          data$CHR == 4 & data$RR == 0|
                                          data$CHR == 6 & data$RR == 0|
                                          data$CHR == 8 & data$RR == 0|
                                          data$CHR == 12 & data$RR == 0, "= 0", "> 0"))

# re shape:
dt <- reshape2::melt(data, id=c("chr", "start", "end" , "CHR", "region", "introgression", "candidate"))

dt <- mutate(dt, Population = ifelse(grepl(glob2rx('*Itabuna*'),dt$variable), "Itabuna",
                                     ifelse(grepl(glob2rx('*Una*'),dt$variable), "Una", "all")))


dt$Population <- factor(dt$Population, levels = c("Itabuna", "Una", "all"))


levels(dt$variable)
```

```{r}
# make plots
p <- dt %>% subset(dt$variable %in% c("Pi_Itabuna", "Pi_Una"))%>%
  ggplot(.,aes(x = candidate, y = log10(value)))+
  facet_wrap(~Population)+
  geom_quasirandom(alpha=.7,size=.3,pch=21,aes(col=Population,fill=Population), width = 1/8)+
  geom_boxplot(aes(fill=Population),alpha=.1,outlier.colour = NA,width=.25,lwd=.1,color="black")+
  stat_compare_means(aes(label = ..p.signif..), label.x = 1.4, label.y = -3)+
  scale_color_manual(values=c("#B79F00", "#821D30"))+
  scale_fill_manual(values=c("#B79F00", "#821D30"))+
  theme_bw()+
  xlab(NULL)+ 
  ylab(expression(paste("log"[10], " nucleotide diversity")))+
  scale_x_discrete(limits=c("= 0", "> 0"), labels = c("LDRs nonrec.", "LDRs rec."))+
  theme(axis.text=element_text(size=7),
        axis.title=element_text(size=8),
        axis.text.x = element_text(angle = 45, hjust = 1),
        legend.position = "none",
        panel.border = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        strip.background = element_blank(),
        plot.title = element_text(size = 9),
        strip.text.x = element_blank(),
        axis.line = element_line(colour = "black")
        )+
  ggtitle("(d)")

```


```{r}
# make final plots for Figure 1:
egg::ggarrange(rrTE, rrIntr,p, nrow = 1, widths = c(.5,.5,1))

```


# Figure 2:

### analyze and plot Nucleotide diversity/Tajima's D

```{r message=FALSE, warning=FALSE}
# load data:
gold <- read_delim("datasets/final.metrics.in250kbwindows.tsv", col_names = T, delim = "\t")


# reshape:
df <- reshape2::melt(gold, id=c("chr", "start", "end", "RR", "TE", "region", "CHR", "introgression"))
df <- mutate(df, Population = ifelse(grepl(glob2rx('*_Itabuna$'),df$variable), "Itabuna",
                                     ifelse(grepl(glob2rx('*_Una$'),df$variable), "Una",
                                            ifelse(grepl(glob2rx('dxy_Una_Taiwan'),df$variable), "Una-Taiwan",
                                                   ifelse(grepl(glob2rx('dxy_Itabuna_Leiden'),df$variable), "Itabuna-Leiden",
                                                          ifelse(grepl(glob2rx('TenUna'),df$variable), "Una-Tenerife",
                                                                 ifelse(grepl(glob2rx('TenIta'),df$variable), "Itabuna-Tenerife","all")))))))

# bin data:
df.bin<-df  %>% subset(!is.na(RR)) %>% mutate( bin.rr = cut( RR, breaks = c(-1,0,5,15,20,200))) 

# add TE island as one bin in the rolling windo RR bin
df.bin$bin.rr<-as.character(df.bin$bin.rr)
df.bin$bin.rr[df.bin$region=="TE islands"]<-"TE islands"
df.bin$bin.rr<-as.factor(df.bin$bin.rr)

#rearrange factor levels
unique(df.bin$bin.rr)
df.bin$bin.rr<-factor(df.bin$bin.rr, levels= c("TE islands","(-1,0]","(0,5]","(5,15]","(15,20]","(20,200]"))

# rearrange and rename Population levels
df.bin$Population<-factor(df.bin$Population,levels= c("Itabuna","Una", "Itabuna-Tenerife", "Una-Tenerife", "Itabuna-Leiden", "Una-Taiwan"),
                          labels=c("Itabuna","Una", "Itabuna-Tenerife", "Una-Tenerife", "Itabuna-Leiden", "Una-Taiwan"))



# correlation between binned RR and genetic diversity:
# filter LGs with large intrgeressions and or having very few localized markers:
data <- subset(df.bin, !(df.bin$CHR %in% c(9,10,11,17,21,26,29)))
data <- data[data$introgression == "no",]

# 
mean(data$RR, na.rm = T)
```

#### Nuc. diversity: column 1
```{r}
options(scipen = -1)

# sub Pi
df.bin.subset <- data %>% subset(RR != "NA" &
                  variable %in% c("Pi_Itabuna", "Pi_Una"))

p1.bin <- df.bin.subset  %>% 
  ggplot(.,aes(x=bin.rr,y=log10(value)))+
  geom_quasirandom(data=subset(df.bin.subset,variable=="Pi_Itabuna"),alpha=.3,size=0.4,stroke=0,aes(col="#B79F00",fill="#B79F00"),pch=21)+
  geom_quasirandom(data=subset(df.bin.subset,variable=="Pi_Una"),alpha=.3,size=0.4,stroke=0,aes(col="#821D30",fill="#821D30"),pch=21)+
  geom_boxplot(aes(fill=region),alpha=.2,outlier.colour = NA,width=.4,lwd=.1,color="black")+
  geom_smooth(data=subset(df.bin.subset,variable=="Pi_Itabuna"),method = "glm", se=T, aes(x=bin.rr,y=log(value,10),group=region),size=.6,alpha=.1,col="#B79F00",fill="#B79F00")+
  geom_smooth(data=subset(df.bin.subset,variable=="Pi_Una"),method = "glm", se=T, aes(x=bin.rr,y=log(value,10),group=region),size=.6,alpha=.1,col="#821D30",fill="#821D30")+
  scale_color_manual(values=c("#821D30", "#B79F00", "#0075DC", "orange"))+
  scale_fill_manual(values=c("#821D30", "#B79F00", "#0075DC", "orange"))+
  scale_x_discrete(name =NULL,
                   labels=c("TEI","0","5","15","20",">20"))+
  coord_cartesian(ylim=c(-5.3,-2))+
  scale_y_continuous(name= expression(paste("log"[10]," nucleotide diversity")), breaks = c(-5,-4,-3))+
  theme_classic2()+
  theme(axis.text = element_text(angle = 0,face="bold",size=10),
        legend.position = "none",
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        panel.border = element_rect(colour = "black", fill=NA, size=1))

# Create a text
grob <- grobTree(textGrob(expression(pi), 
                          gp=gpar(col="grey", fontsize=80, fontface="bold",alpha=.4, fontfamily = "Georgia")))
# Plot
p1.bin<-p1.bin + annotation_custom(grob)

# pval
grob <- grobTree(textGrob(expression(paste("Itabuna: ",italic(rho)," = 0.28****")), x=0.2,  y=0.95, hjust=0,
                          gp=gpar(col="#B79F00", fontsize=8, fontface="bold",alpha=1)))
p1.bin<-p1.bin + annotation_custom(grob)

grob <- grobTree(textGrob(expression(paste("Una: ",italic(rho)," = 0.25****")), x=0.2,  y=0.88, hjust=0,
                          gp=gpar(col="#821D30", fontsize=8, fontface="bold",alpha=1)))
p1.bin<-p1.bin + annotation_custom(grob)


# sub Tajima's D
df.bin.subset <- data %>% subset(RR != "NA" &
                                   variable %in% c("tajD_Itabuna", "tajD_Una"))



t1.bin <- df.bin.subset  %>% 
  ggplot(.,aes(x=bin.rr,y=value))+
  geom_quasirandom(data=subset(df.bin.subset,variable=="tajD_Itabuna"),alpha=.3,size=0.4,stroke=0,aes(col="#B79F00",fill="#B79F00"),pch=21)+
  geom_quasirandom(data=subset(df.bin.subset,variable=="tajD_Una"),alpha=.3,size=0.4,stroke=0,aes(col="#821D30",fill="#821D30"),pch=21)+
  geom_boxplot(aes(fill=region),alpha=.2,outlier.colour = NA,width=.4,lwd=.1,color="black")+
  geom_smooth(data=subset(df.bin.subset,variable=="tajD_Itabuna"),method = "glm", se=T, aes(x=bin.rr,y=value,group=region),size=.6,alpha=.1,col="#B79F00",fill="#B79F00")+
  geom_smooth(data=subset(df.bin.subset,variable=="tajD_Una"),method = "glm", se=T, aes(x=bin.rr,y=value,group=region),size=.6,alpha=.1,col="#821D30",fill="#821D30")+
  scale_color_manual(values=c("#821D30", "#B79F00", "#0075DC", "orange"))+
  scale_fill_manual(values=c("#821D30", "#B79F00", "#0075DC", "orange"))+
  scale_x_discrete(name =NULL,
                   labels=c("TEI","0","5","15","20",">20"))+
                   # labels=c("TEI","0","5","10","15","20",">20"))+
  scale_y_continuous(name= expression(paste("Tajima's ", italic(D))), breaks = c(-2,-1,0,1,2))+
  theme_classic2()+
  coord_cartesian(ylim=c(-2.8,3))+
  theme(axis.text = element_text(angle = 0,face="bold",size=10),
        legend.position = "none",
        panel.border = element_rect(colour = "black", fill=NA, size=1))



# Create a text
grob <- grobTree(textGrob(expression(paste("Taj", italic(D))),
                          gp=gpar(col="grey", fontsize=60, fontface="bold",alpha=.4, fontfamily = "Georgia")))
# Plot
t1.bin<-t1.bin + annotation_custom(grob)

# pval
grob <- grobTree(textGrob(expression(paste("Itabuna: ",italic(rho)," = 0.21****")), x=0.2,  y=0.95, hjust=0,
                          gp=gpar(col="#B79F00", fontsize=8, fontface="bold",alpha=1)))
t1.bin<-t1.bin + annotation_custom(grob)

grob <- grobTree(textGrob(expression(paste("Una: ",italic(rho)," = 0.30****")), x=0.2,  y=0.88, hjust=0,
                          gp=gpar(col="#821D30", fontsize=8, fontface="bold",alpha=1)))
t1.bin<-t1.bin + annotation_custom(grob)


egg::ggarrange(p1.bin,t1.bin, nrow = 2)

```
#### Differentiation/divergence in LDRs: column 2
```{r}
# sub Fst
df.bin.subset <- data %>% subset(RR != "NA" & 
                                   variable %in% c("Fst_Tenerife_Itabuna", "Fst_Tenerife_Una"))


f1.bin <- df.bin.subset  %>% 
  ggplot(.,aes(x=bin.rr,y=value))+
  geom_quasirandom(data=subset(df.bin.subset,variable=="Fst_Tenerife_Itabuna"),alpha=.3,size=0.4,stroke=0,aes(col="#B79F00",fill="#B79F00"),pch=21)+
  geom_quasirandom(data=subset(df.bin.subset,variable=="Fst_Tenerife_Una"),alpha=.3,size=0.4,stroke=0,aes(col="#821D30",fill="#821D30"),pch=21)+
  geom_boxplot(aes(fill=region),alpha=.2,outlier.colour = NA,width=.4,lwd=.1,color="black")+
  geom_smooth(data=subset(df.bin.subset,variable=="Fst_Tenerife_Itabuna"),method = "glm", se=T, aes(x=bin.rr,y=value,group=region),size=.6,alpha=.1,col="#B79F00",fill="#B79F00")+
  geom_smooth(data=subset(df.bin.subset,variable=="Fst_Tenerife_Una"),method = "glm", se=T, aes(x=bin.rr,y=value,group=region),size=.6,alpha=.1,col="#821D30",fill="#821D30")+
  scale_color_manual(values=c("#821D30", "#B79F00", "#0075DC", "orange"))+
  scale_fill_manual(values=c("#821D30", "#B79F00", "#0075DC", "orange"))+
  scale_x_discrete(name =NULL,
                   labels=c("TEI","0","5","15","20",">20"))+
                   # labels=c("TEI","0","5","10","15","20",">20"))+
  coord_cartesian(ylim=c(0.2,1.1))+
  scale_y_continuous(name= expression(italic("F")["ST"]), breaks = c(0,.3,.6,.9))+
  theme_classic2()+
  theme(axis.text = element_text(angle = 0,face="bold",size=10),
        legend.position = "none",
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        panel.border = element_rect(colour = "black", fill=NA, size=1))

# Create a text
grob <- grobTree(textGrob(expression(paste(italic(F)[ST])), 
                          gp=gpar(col="grey", fontsize=60, fontface="bold",alpha=.4, fontfamily = "Georgia")))
# Plot
f1.bin<-f1.bin + annotation_custom(grob)

# pval
grob <- grobTree(textGrob(expression(paste("Itabuna-Tenerife: ",italic(rho)," = -0.18***")), x=0.165,  y=0.95, hjust=0,
                          gp=gpar(col="#B79F00", fontsize=8, fontface="bold",alpha=1)))
f1.bin<-f1.bin + annotation_custom(grob)

grob <- grobTree(textGrob(expression(paste("Una-Tenerife: ",italic(rho)," = -0.12**")), x=0.165,  y=0.88, hjust=0,
                          gp=gpar(col="#821D30", fontsize=8, fontface="bold",alpha=1)))
f1.bin<-f1.bin + annotation_custom(grob)



# sub Dxy
df.bin.subset <- data %>% subset(RR != "NA" &
                                   variable %in% c("dxy_Itabuna_Leiden", "dxy_Una_Taiwan"))

options(scipen = 1)


d1.bin <- df.bin.subset  %>% 
  ggplot(.,aes(x=bin.rr,y=log10(value)))+
  geom_quasirandom(data=subset(df.bin.subset,variable=="dxy_Itabuna_Leiden"),alpha=.3,size=0.4,stroke=0,aes(col="#B79F00",fill="#B79F00"),pch=21)+
  geom_quasirandom(data=subset(df.bin.subset,variable=="dxy_Una_Taiwan"),alpha=.3,size=0.4,stroke=0,aes(col="#821D30",fill="#821D30"),pch=21)+
  geom_boxplot(aes(fill=region),alpha=.2,outlier.colour = NA,width=.4,lwd=.1,color="black")+
  geom_smooth(data=subset(df.bin.subset,variable=="dxy_Itabuna_Leiden"),method = "glm", se=T, aes(x=bin.rr,y=log10(value),group=region),size=.6,alpha=.1,col="#B79F00",fill="#B79F00")+
  geom_smooth(data=subset(df.bin.subset,variable=="dxy_Una_Taiwan"),method = "glm", se=T, aes(x=bin.rr,y=log10(value),group=region),size=.6,alpha=.1,col="#821D30",fill="#821D30")+
  scale_color_manual(values=c("#821D30", "#B79F00", "#0075DC", "orange"))+
  scale_fill_manual(values=c("#821D30", "#B79F00", "#0075DC", "orange"))+
  scale_x_discrete(name =NULL,
                   labels=c("TEI","0","5","15","20",">20"))+
                   # labels=c("TEI","0","5","10","15","20",">20"))+
  coord_cartesian(ylim=c(-3.5,-1.8))+
  scale_y_continuous(name= expression(paste("log"[10]," ",italic("d")["xy"])), breaks = c(-3.5,-3.0,-2.5,-2))+
  theme_classic2()+
  theme(axis.text = element_text(angle = 0,face="bold",size=10),
        legend.position = "none",
        panel.border = element_rect(colour = "black", fill=NA, size=1))


# Create a text
grob <- grobTree(textGrob(expression(paste(italic(d)[xy])),
                          gp=gpar(col="grey", fontsize=65, fontface="bold",alpha=.4, fontfamily = "Georgia")))
# Plot
d1.bin<-d1.bin + annotation_custom(grob)

# pval
grob <- grobTree(textGrob(expression(paste("Itabuna-Leiden: ",italic(rho)," = 0.41****")), x=0.165,  y=0.95, hjust=0,
                          gp=gpar(col="#B79F00", fontsize=8, fontface="bold",alpha=1)))
d1.bin<-d1.bin + annotation_custom(grob)

grob <- grobTree(textGrob(expression(paste("Una-Taiwan: ",italic(rho)," = 0.37****")), x=0.165,  y=0.88, hjust=0,
                          gp=gpar(col="#821D30", fontsize=8, fontface="bold",alpha=1)))
d1.bin<-d1.bin + annotation_custom(grob)


egg::ggarrange(f1.bin,d1.bin, nrow = 2)

```

Nuc. diversity at syn/nonsyn positions: column 3

```{r}
# load data:
gold <- read_delim("datasets/final.metrics.in250kbwindows.tsv", col_names = T, delim = "\t")

# reshape:
df <- reshape2::melt(gold[,c(1:7,15:18,26)], id=c("chr", "start", "end", "RR", "TE", "region", "CHR", "introgression"))
df <- mutate(df, Population = ifelse(grepl(glob2rx('*_Itabuna$'),df$variable), "Itabuna",
                                     ifelse(grepl(glob2rx('*_Una$'),df$variable), "Una","all")))

# bin data:
df.bin<-df  %>% subset(!is.na(RR)) %>% mutate( bin.rr = cut( RR, breaks = c(-1,0,5,15,20,200))) 


# add TE island as one bin in the rolling windo RR bin
df.bin$bin.rr<-as.character(df.bin$bin.rr)
df.bin$bin.rr[df.bin$region=="TE islands"]<-"TE islands"
df.bin$bin.rr<-as.factor(df.bin$bin.rr)

#rearrange factor levels
unique(df.bin$bin.rr)
df.bin$bin.rr<-factor(df.bin$bin.rr, levels= c("TE islands","(-1,0]","(0,5]","(5,15]","(15,20]","(20,200]"))

# rearrange and rename Population levels
df.bin$Population<-factor(df.bin$Population,levels= c("Itabuna","Una"),
                          labels=c("Itabuna","Una"))



# correlation between binned RR and genetic diversity:
# filter LGs with large intrgeressions and or having very few localized markers:
data <- subset(df.bin, !(df.bin$CHR %in% c(9,10,11,17,21,26,29)))
data <- data[data$introgression == "no",]
# Exclude LDRs for diversity
# data <- data[data$region == "LDRs",]

# 
mean(data$RR, na.rm = T)
```

```{r}
# sub Pi syn
df.bin.subset <- data %>% subset(RR != "NA" &
                                   variable %in% c("Pi_syn_Itabuna", "Pi_syn_Una"))



ps1.bin <- df.bin.subset  %>% 
  ggplot(.,aes(x=bin.rr,y=log10(value)))+
  geom_quasirandom(data=subset(df.bin.subset,variable=="Pi_syn_Itabuna"),alpha=.3,size=0.4,stroke=0,aes(col="#B79F00",fill="#B79F00"),pch=21)+
  geom_quasirandom(data=subset(df.bin.subset,variable=="Pi_syn_Una"),alpha=.3,size=0.4,stroke=0,aes(col="#821D30",fill="#821D30"),pch=21)+
  geom_boxplot(aes(fill=region),alpha=.2,outlier.colour = NA,width=.4,lwd=.1,color="black")+
  geom_smooth(data=subset(df.bin.subset,variable=="Pi_syn_Itabuna"),method = "glm", se=T, aes(x=bin.rr,y=log10(value),group=region),size=.6,alpha=.1,col="#B79F00",fill="#B79F00")+
  geom_smooth(data=subset(df.bin.subset,variable=="Pi_syn_Una"),method = "glm", se=T, aes(x=bin.rr,y=log10(value),group=region),size=.6,alpha=.1,col="#821D30",fill="#821D30")+
  scale_color_manual(values=c("#821D30", "#B79F00", "#0075DC", "orange"))+
  scale_fill_manual(values=c("#821D30", "#B79F00", "#0075DC", "orange"))+
  scale_x_discrete(name =NULL,
                   labels=c("TEI","0","5","15","20",">20"))+
                   # labels=c("TEI","0","5","10","15","20",">20"))+
  coord_cartesian(ylim=c(-5.4,-2))+
  scale_y_continuous(name= expression(paste("log"[10]," ",pi["S"])), breaks = c(-5,-4,-3))+
  theme_classic2()+
  theme(axis.text = element_text(angle = 0,face="bold",size=10),
        legend.position = "none",
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        panel.border = element_rect(colour = "black", fill=NA, size=1))


# Create a text
grob <- grobTree(textGrob(expression(paste(pi)[S]),
                          gp=gpar(col="grey", fontsize=80, fontface="bold",alpha=.4, fontfamily = "Georgia")))
# Plot
ps1.bin<-ps1.bin + annotation_custom(grob)


# pval
grob <- grobTree(textGrob(expression(paste("Itabuna: ",italic(rho)," = 0.26****")), x=0.2,  y=0.95, hjust=0,
                          gp=gpar(col="#B79F00", fontsize=8, fontface="bold",alpha=1)))
ps1.bin<-ps1.bin + annotation_custom(grob)

grob <- grobTree(textGrob(expression(paste("Una: ",italic(rho)," = 0.22****")), x=0.2,  y=0.88, hjust=0,
                          gp=gpar(col="#821D30", fontsize=8, fontface="bold",alpha=1)))
ps1.bin<-ps1.bin + annotation_custom(grob)



# sub Pi nonsyn
df.bin.subset <- data %>% subset(RR != "NA" & 
                                   variable %in% c("Pi_nonsyn_Itabuna", "Pi_nonsyn_Una"))




pns1.bin <- df.bin.subset  %>% 
  ggplot(.,aes(x=bin.rr,y=log10(value)))+
  geom_quasirandom(data=subset(df.bin.subset,variable=="Pi_nonsyn_Itabuna"),alpha=.3,size=0.4,stroke=0,aes(col="#B79F00",fill="#B79F00"),pch=21)+
  geom_quasirandom(data=subset(df.bin.subset,variable=="Pi_nonsyn_Una"),alpha=.3,size=0.4,stroke=0,aes(col="#821D30",fill="#821D30"),pch=21)+
  geom_boxplot(aes(fill=region),alpha=.2,outlier.colour = NA,width=.4,lwd=.1,color="black")+
  geom_smooth(data=subset(df.bin.subset,variable=="Pi_nonsyn_Itabuna"),method = "glm", se=T, aes(x=bin.rr,y=log10(value),group=region),size=.6,alpha=.1,col="#B79F00",fill="#B79F00")+
  geom_smooth(data=subset(df.bin.subset,variable=="Pi_nonsyn_Una"),method = "glm", se=T, aes(x=bin.rr,y=log10(value),group=region),size=.6,alpha=.1,col="#821D30",fill="#821D30")+
  scale_color_manual(values=c("#821D30", "#B79F00", "#0075DC", "orange"))+
  scale_fill_manual(values=c("#821D30", "#B79F00", "#0075DC", "orange"))+
  scale_x_discrete(name =NULL,
                   labels=c("TEI","0","5","15","20",">20"))+
                   # labels=c("TEI","0","5","10","15","20",">20"))+
  coord_cartesian(ylim=c(-5.8,-2))+
  scale_y_continuous(name= expression(paste("log"[10]," ",pi["NS"])), breaks = c(-5,-4,-3))+
  theme_classic2()+
  theme(axis.text = element_text(angle = 0,face="bold",size=10),
        legend.position = "none",
        panel.border = element_rect(colour = "black", fill=NA, size=1))


# Create a text
grob <- grobTree(textGrob(expression(paste(pi)[NS]),
                          gp=gpar(col="grey", fontsize=80, fontface="bold",alpha=.4, fontfamily = "Georgia")))
# Plot
pns1.bin<-pns1.bin + annotation_custom(grob)


# pval
grob <- grobTree(textGrob(expression(paste("Itabuna: ",italic(rho)," = 0.21****")), x=0.2,  y=0.95, hjust=0,
                          gp=gpar(col="#B79F00", fontsize=8, fontface="bold",alpha=1)))
pns1.bin<-pns1.bin + annotation_custom(grob)

grob <- grobTree(textGrob(expression(paste("Una: ",italic(rho)," = 0.22****")), x=0.2,  y=0.88, hjust=0,
                          gp=gpar(col="#821D30", fontsize=8, fontface="bold",alpha=1)))
pns1.bin<-pns1.bin + annotation_custom(grob)


annotate_figure(ggpubr::ggarrange(egg::ggarrange(p1.bin,t1.bin, nrow = 2),
                                  egg::ggarrange(ps1.bin,pns1.bin, nrow = 2),
                                  egg::ggarrange(f1.bin,d1.bin, nrow = 2), nrow = 1, ncol = 3),
                bottom = text_grob("binned recombination rate (cM/Mb)", size = 10))

# library(extrafont)
# 
# loadfonts(device = "all")


dev.print(pdf,"~/sciebo/My_PhD/Cobs3.1/linkageMap/manuscript/revisions/MainMaterial/binnedRR.Pi.TajD.Fst.Dxy.Pisyn.Pinonsyn.fig2_v6.pdf",width=8,height=6)

```
# Figure 3:

### analyze and plot SVs and InDels:
analyze associations between recombination and structural variants.

Load good stuffs for plotting/analyzing
```{r}
# load data:
sv <- read_delim("datasets/SVs/sv.type.recomb.repeat.freq.all.txt", col_names = T, delim = "\t")

# annotate TE islands:
sv <- mutate(sv, region=ifelse(sv$chr == "LG1" & 3750000 <= sv$start & sv$end <= 4500000|
                                 sv$chr == "LG1" & 13750000 <= sv$start & sv$end <= 13819712|
                                 sv$chr == "LG2" & 4250000 <= sv$start & sv$end <= 6250000|
                                 sv$chr == "LG3" & 5750000 <= sv$start & sv$end <= 8000000|
                                 sv$chr == "LG4" & 0 <= sv$start & sv$end <= 750000|
                                 sv$chr == "LG5" & 0 <= sv$start & sv$end <= 250000|
                                 sv$chr == "LG5" & 3750000 <= sv$start & sv$end <= 5000000|
                                 sv$chr == "LG7" & 3750000 <= sv$start & sv$end <= 6000000|
                                 sv$chr == "LG8" & 0 <= sv$start & sv$end <= 1000000|
                                 sv$chr == "LG9" & 0 <= sv$start & sv$end <= 2000000|
                                 sv$chr == "LG10" & 0 <= sv$start & sv$end <= 1000000|
                                 sv$chr == "LG10" & 4000000 <= sv$start & sv$end <= 4750000|
                                 sv$chr == "LG11" & 6250000 <= sv$start & sv$end <= 7361187|
                                 sv$chr == "LG13" & 0 <= sv$start & sv$end <= 1500000|
                                 sv$chr == "LG14" & 3500000 <= sv$start & sv$end <= 5750000|
                                 sv$chr == "LG15" & 5250000 <= sv$start & sv$end <= 5631722|
                                 sv$chr == "LG16" & 4250000 <= sv$start & sv$end <= 5477151|
                                 sv$chr == "LG17" & 4500000 <= sv$start & sv$end <= 5213470|
                                 sv$chr == "LG19" & 4000000 <= sv$start & sv$end <= 5027330|
                                 sv$chr == "LG20" & 0 <= sv$start & sv$end <= 1250000|
                                 sv$chr == "LG21" & 2500000 <= sv$start & sv$end <= 4437275|
                                 sv$chr == "LG22" & 0 <= sv$start & sv$end <= 1000000|
                                 sv$chr == "LG22" & 4000000 <= sv$start & sv$end <= 4343458|
                                 sv$chr == "LG23" & 0 <= sv$start & sv$end <= 1000000|
                                 sv$chr == "LG24" & 3250000 <= sv$start & sv$end <= 3702185|
                                 sv$chr == "LG25" & 1750000 <= sv$start & sv$end <= 3136423|
                                 sv$chr == "LG26" & 1500000 <= sv$start & sv$end <= 2304265|
                                 sv$chr == "LG27" & 0 <= sv$start & sv$end <= 750000|
                                 sv$chr == "LG28" & 0 <= sv$start & sv$end <= 500000|
                                 sv$chr == "LG29" & 0 <= sv$start & sv$end <= 500000,
                               "TE islands", "LDRs"))



# annotate introgressed regions:
sv <- mutate(sv, introgression=ifelse(sv$chr == "LG1" & 500000 <= sv$start & sv$end <= 1250000|
                                          sv$chr == "LG10" & 0 <= sv$start & sv$end <= 5000000|
                                          sv$chr == "LG10" & 6500000 <= sv$start & sv$end <= 7000000|
                                          sv$chr == "LG11" & 0 <= sv$start & sv$end <= 250000|
                                          sv$chr == "LG11" & 3750000 <= sv$start & sv$end <= 4000000|
                                          sv$chr == "LG11" & 4250000 <= sv$start & sv$end <= 5000000|
                                          sv$chr == "LG11" & 5250000 <= sv$start & sv$end <= 5750000|
                                          sv$chr == "LG15" & 4750000 <= sv$start & sv$end <= 5000000|
                                          sv$chr == "LG17" & 0 <= sv$start & sv$end <= 750000|
                                          sv$chr == "LG17" & 1750000 <= sv$start & sv$end <= 2000000|
                                          sv$chr == "LG17" & 3750000 <= sv$start & sv$end <= 5250000|
                                          sv$chr == "LG2" & 2250000 <= sv$start & sv$end <= 3250000|
                                          sv$chr == "LG20" & 2750000 <= sv$start & sv$end <= 3000000|
                                          sv$chr == "LG21" & 2250000 <= sv$start & sv$end <= 4250000|
                                          sv$chr == "LG23" & 2750000 <= sv$start & sv$end <= 3500000|
                                          sv$chr == "LG3" & 0 <= sv$start & sv$end <= 500000|
                                          sv$chr == "LG3" & 10500000 <= sv$start & sv$end <= 10750000|
                                          sv$chr == "LG4" & 9500000 <= sv$start & sv$end <= 9750000|
                                          sv$chr == "LG9" & 3500000 <= sv$start & sv$end <= 6500000,
                                        "yes", "no"))


sv$CHR <- as.numeric(gsub("\\D", "", sv$chr))

```

plot corr with stats:
```{r}
# Analyse correlations:
# data <- sv
data <- subset(sv, !(sv$CHR %in% c(9,10,11,17,21,26,29)))
data <- data[data$introgression == "no",]
data <- data[data$meanperc < 80,]

# Exclude LDRs for diversity
data <- data[data$region == "LDRs",]

# How many SVs? 481
ct <- na.omit(data)

# re shape:
dt <- reshape2::melt(data, id=c("chr", "start", "end" , "type", "window", "CHR", "id", "meanperc", "region", "introgression", "recomb",  "Ref_Allele_all"))

```

Make two panels of Figure 3
```{r}

# make binned RR plots
data <- subset(sv, !(sv$CHR %in% c(9,10,11,17,21,26,29)))
data <- data[data$introgression == "no",]
data <- data[data$meanperc < 80,]

# Exclude TE islands
data <- data[data$region == "LDRs",]

# bin data:
df.bin<- data  %>% subset(!is.na(recomb)) %>% mutate( bin.rr = cut( recomb, breaks = c(-1,0,5,15,20,200))) 

# add TE island as one bin in the rolling windo RR bin
df.bin$bin.rr<-as.character(df.bin$bin.rr)
df.bin$bin.rr<-as.factor(df.bin$bin.rr)

#rearrange factor levels
unique(df.bin$bin.rr)
df.bin$bin.rr<-factor(df.bin$bin.rr, levels= c("(-1,0]","(0,5]","(5,15]","(15,20]","(20,200]"))


# correlation between binned RR and genome features data:
l<-df.bin %>% 
  ggplot(.,aes(x=bin.rr,y=log10(length)))+
  geom_quasirandom(alpha=.4,size=0.6,stroke=0,aes(col=type,fill=type),pch=21)+
  geom_boxplot(aes(),alpha=.2,outlier.colour = NA,width=.4,lwd=.1,color="black", fill="#0075DC")+
  geom_smooth(data=subset(df.bin,region=="LDRs"),method = "glm", se=T, aes(x=bin.rr,y=log(length,10),group=region),size=1,alpha=.1,col=rgb(1,0,0,.8),fill="red2")+
  scale_color_manual(values=c("#76B947","#B1D8B7","#2F5233"))+
  scale_fill_manual(values=c("#76B947","#B1D8B7","#2F5233"))+
  scale_x_discrete(name =NULL, 
                   labels=c("0","5","15","20",">20"))+
  scale_y_continuous(name=expression(paste("SV length [log"[10], " bp]")))+ #, breaks = c(-5,-4,-3)
  theme_classic2()+
  theme(axis.text = element_text(angle = 0,face="bold",size=10),
        legend.position = "none",
        panel.border = element_rect(colour = "black", fill=NA, size=1)) + ggtitle("(a)")


# Create a text
grob <- grobTree(textGrob(expression(paste("SV: ",italic(rho)," = 0.14**")), x=0.6,  y=0.95, hjust=0,
                          gp=gpar(col="red2", fontsize=8, fontface="bold",alpha=1)))
# Plot
l<-l + annotation_custom(grob)

grob <- grobTree(textGrob(expression(paste("DEL: ",italic(rho)," = 0.13")), x=0.6,  y=0.88, hjust=0,
                          gp=gpar(col="#76B947", fontsize=8, fontface="bold",alpha=1)))
# Plot
l<-l + annotation_custom(grob)

grob <- grobTree(textGrob(expression(paste("DUP: ",italic(rho)," = -0.08")), x=0.6,  y=0.81, hjust=0,
                          gp=gpar(col="#B1D8B7", fontsize=8, fontface="bold",alpha=1)))
l<-l + annotation_custom(grob)

grob <- grobTree(textGrob(expression(paste("INV: ",italic(rho)," = 0.05")), x=0.6,  y=0.74, hjust=0,
                          gp=gpar(col="#2F5233", fontsize=8, fontface="bold",alpha=1)))
l<-l + annotation_custom(grob)



fa<-df.bin %>% 
  ggplot(.,aes(x=bin.rr,y=Alt_Allele_all))+
  geom_quasirandom(alpha=.4,size=0.6,stroke=0,aes(col=type,fill=type),pch=21)+
  geom_boxplot(aes(),alpha=.2,outlier.colour = NA,width=.4,lwd=.1,color="black", fill="#0075DC")+
  geom_smooth(data=subset(df.bin,region=="LDRs"),method = "glm", se=T, aes(x=bin.rr,y=Alt_Allele_all,group=region),size=1,alpha=.1,col=rgb(1,0,0,.8),fill="red2")+
  scale_color_manual(values=c("#76B947","#B1D8B7","#2F5233"))+
  scale_fill_manual(values=c("#76B947","#B1D8B7","#2F5233"))+
  scale_x_discrete(name =NULL, 
                   labels=c("0","5","15","20",">20"))+
  scale_y_continuous(name="SV frequency")+ #, breaks = c(-5,-4,-3)
  theme_classic2()+
  theme(axis.text = element_text(angle = 0,face="bold",size=10),
        legend.position = "none",
        panel.border = element_rect(colour = "black", fill=NA, size=1)) + ggtitle("(b)")


# Create a text
grob <- grobTree(textGrob(expression(paste("SV: ",italic(rho)," = 0.15***")), x=0.6,  y=0.95, hjust=0,
                          gp=gpar(col="red2", fontsize=8, fontface="bold",alpha=1)))
# Plot
fa<-fa + annotation_custom(grob)

grob <- grobTree(textGrob(expression(paste("DEL: ",italic(rho)," = 0.06")), x=0.6,  y=0.88, hjust=0,
                          gp=gpar(col="#76B947", fontsize=8, fontface="bold",alpha=1)))
# Plot
fa<-fa + annotation_custom(grob)

grob <- grobTree(textGrob(expression(paste("DUP: ",italic(rho)," = 0.13")), x=0.6,  y=0.81, hjust=0,
                          gp=gpar(col="#B1D8B7", fontsize=8, fontface="bold",alpha=1)))
fa<-fa + annotation_custom(grob)

grob <- grobTree(textGrob(expression(paste("INV: ",italic(rho)," = 0.06")), x=0.6,  y=0.74, hjust=0,
                          gp=gpar(col="#2F5233", fontsize=8, fontface="bold",alpha=1)))
fa<-fa + annotation_custom(grob)


egg::ggarrange(l, fa,  nrow = 1)

```


Load polymorphism data at indels to create panel 3 of Figure 3
```{r}
# load data:
gold <- read_delim("datasets/final.metrics.in250kbwindows.tsv", col_names = T, delim = "\t")


# reshape:
df <- reshape2::melt(gold[,c(1:7,14,26)], id=c("chr", "start", "end", "RR", "TE", "region", "CHR", "introgression"))


# bin data:
df.bin<-df  %>% subset(!is.na(RR)) %>% mutate( bin.rr = cut( RR, breaks = c(-1,0,5,15,20,200)))

# add TE island as one bin in the rolling windo RR bin
df.bin$bin.rr<-as.character(df.bin$bin.rr)
df.bin$bin.rr[df.bin$region=="TE islands"]<-"TE islands"
df.bin$bin.rr<-as.factor(df.bin$bin.rr)

#rearrange factor levels
unique(df.bin$bin.rr)
df.bin$bin.rr<-factor(df.bin$bin.rr, levels= c("TE islands","(-1,0]","(0,5]","(5,15]","(15,20]","(20,200]"))

# rearrange and rename Population levels
df.bin$Population<-factor(df.bin$variable,levels= c("pi_indels"), labels=c("pi_indels"))



# correlation between binned RR and genetic diversity:
# filter LGs with large intrgeressions and or having very few localized markers:
data <- subset(df.bin, !(df.bin$CHR %in% c(9,10,11,17,21,26,29)))
data <- data[data$introgression == "no",]
# 
mean(data$RR, na.rm = T)
```


Make final Figure 3
```{r}
options(scipen = 1)

# sub Pi
df.bin.subset <- data %>% subset(RR != "NA" & region != "TE islands" & 
                  variable %in% c("pi_indels"))


p1.bin <-   df.bin.subset  %>% 
  ggplot(.,aes(x=bin.rr,y=log10(value)))+
  geom_quasirandom(alpha=.4,size=0.6,stroke=0,aes(col=region,fill=region),pch=21)+
  geom_boxplot(aes(fill=region),alpha=.2,outlier.colour = NA,width=.4,lwd=.1,color="black")+
  geom_smooth(data=subset(df.bin.subset,region=="LDRs"),method = "glm", se=T, 
              aes(x=bin.rr,y=log(value,10),group=region),size=1,alpha=.1,col=rgb(1,0,0,.8),fill="red2")+
  scale_color_manual(values=c("#0075DC","orange"))+
  scale_fill_manual(values=c("#0075DC","orange"))+
  scale_x_discrete(name =NULL, 
                   labels=c("0","5","15","20",">20"))+
  scale_y_continuous(name=expression(paste("log"[10]," ",pi[indels])), breaks = c(-4.5,-4.0,-3.5))+
  theme_classic2()+
  theme(axis.text = element_text(angle = 0,face="bold",size=10),
        legend.position = "none",
        panel.border = element_rect(colour = "black", fill=NA, size=1)) + ggtitle("(c)")

# Create a text
grob <- grobTree(textGrob(expression(paste(italic(rho)," = 0.54****")), x=0.6,  y=0.95, hjust=0,
                          gp=gpar(col="red2", fontsize=8, fontface="bold",alpha=1)))
# Plot
p1.bin<-p1.bin + annotation_custom(grob)


annotate_figure(egg::ggarrange(l, fa, p1.bin, nrow = 1), bottom = text_grob("binned recombination rate (cM/Mb)", size = 10))
```